pkg/controller/certificatesigningrequests/venafi/venafi.go

```diff
@@ -0,0 +1,206 @@
+	controllerpkg "github.com/jetstack/cert-manager/pkg/controller"
+	"github.com/jetstack/cert-manager/pkg/controller/certificatesigningrequests"
+	"github.com/jetstack/cert-manager/pkg/controller/certificatesigningrequests/util"
+	venaficlient "github.com/jetstack/cert-manager/pkg/issuer/venafi/client"
+	venafiapi "github.com/jetstack/cert-manager/pkg/issuer/venafi/client/api"
+	logf "github.com/jetstack/cert-manager/pkg/logs"
+	"github.com/jetstack/cert-manager/pkg/util/pki"
+	utilpki "github.com/jetstack/cert-manager/pkg/util/pki"
+)
+
+const (
+	CSRControllerName = "certificatesigningrequests-issuer-venafi"
+)
+
+// Venafi is a Kubernetes CertificateSigningRequest controller, responsible for
+// signing CertificateSigningRequests that reference a cert-manager Venafi
+// Issuer or ClusterIssuer
+type Venafi struct {
+	issuerOptions controllerpkg.IssuerOptions
+	secretsLister corelisters.SecretLister
+	certClient    certificatesclient.CertificateSigningRequestInterface
+	recorder      record.EventRecorder
+
+	clientBuilder venaficlient.VenafiClientBuilder
+}
+
+func init() {
+	controllerpkg.Register(CSRControllerName, func(ctx *controllerpkg.Context) (controllerpkg.Interface, error) {
+		return controllerpkg.NewBuilder(ctx, CSRControllerName).
+			For(certificatesigningrequests.New(apiutil.IssuerVenafi, NewVenafi(ctx))).
+			Complete()
+	})
+}
+
+func NewVenafi(ctx *controllerpkg.Context) *Venafi {
+	return &Venafi{
+		issuerOptions: ctx.IssuerOptions,
+		secretsLister: ctx.KubeSharedInformerFactory.Core().V1().Secrets().Lister(),
+		certClient:    ctx.Client.CertificatesV1().CertificateSigningRequests(),
+		recorder:      ctx.Recorder,
+		clientBuilder: venaficlient.New,
+	}
+}
+
+// Sign attempts to sign the given CertificateSigningRequest based on the
+// provided Venafi Issuer or ClusterIssuer. This function will update the resource
+// if signing was successful. Returns an error which, if not nil, should
+// trigger a retry.
+// Since this signer takes some time to sign the request, this controller will
+// set a "pick ID" annotation value that is used to fetch the latest state of
+// the request in subsequent re-syncs.
+func (v *Venafi) Sign(ctx context.Context, csr *certificatesv1.CertificateSigningRequest, issuerObj cmapi.GenericIssuer) error {
+	log := logf.FromContext(ctx, "sign")
+	log = logf.WithRelatedResource(log, issuerObj)
+
+	resourceNamespace := v.issuerOptions.ResourceNamespace(issuerObj)
+
+	client, err := v.clientBuilder(resourceNamespace, v.secretsLister, issuerObj)
+	if apierrors.IsNotFound(err) {
+		message := "Required secret resource not found"
+		v.recorder.Event(csr, corev1.EventTypeWarning, "SecretNotFound", message)
+		log.Error(err, message)
+		return nil
+	}
+
+	if err != nil {
+		message := fmt.Sprintf("Failed to initialise venafi client for signing: %s", err)
+		v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorVenafiInit", message)
+		log.Error(err, message)
+		return err
```

Comment:
question: are the errors returned by `clientBuilder` retryiable? If they are, where is that documented? Looking at [New](https://github.com/jetstack/cert-manager/blob/dffbf391dbb0fc6c1cfea62e561a9c6f54362ab0/pkg/issuer/venafi/client/venaficlient.go#L73), [vcert.NewClient](https://github.com/Venafi/vcert/blob/f1b4862abb7b800ac9558ad9b9c201dca6261b43/client.go#L34-L64) does not seem to be doing any network call. I'd thus not return any error here.

Would it be possible to document `New`?

Reply:
That function has retryiable errors. It does both lookups to resources in [kube](https://github.com/jetstack/cert-manager/blob/dffbf391dbb0fc6c1cfea62e561a9c6f54362ab0/pkg/issuer/venafi/client/venaficlient.go#L98), as well as network calls to [Venafi](https://github.com/Venafi/vcert/blob/f1b4862abb7b800ac9558ad9b9c201dca6261b43/pkg/venafi/cloud/connector.go#L348).

I will add a comment to `New` that errors may be network errors.

Reply:
> It does both lookups to resources in kube, as well as network calls to Venafi.

You are right, it does look up for secrets (I missed the API call hidden in `configForIssuer`).

That said, I can't see any retriable failure coming from `vcert.NewClient`.

Reply:
From here: https://github.com/Venafi/vcert/blob/4c2b2399c2960e79f8b5708798cc1dd3b4f5db69/client.go#L62